improve multi-platform build and frontend performance

imenattatra opened the pull request:
## AI Disclaimer: 
Gemini was used

## What type of PR is this? 

- [x] Refactor


## Summary / Code Strategy

This PR refactors the Dockerfile to enhance multi-platform build support and improve frontend build performance. By leveraging native platform builds and optimizing Yarn configuration, the changes streamline the build process, particularly for ARM-based and multi-architecture environments. 

## Description

- Add `BUILDPLATFORM` and `TARGETPLATFORM` build arguments and use` --platform=$BUILDPLATFORM` for the frontend builder stage
to enable native platform builds (improves speed, especially on ARM Macs).
- Remove `--network-concurrency 1` from Yarn install, allowing parallel downloads now that native builds are used.
- Minor cleanup and reordering for clarity.

These changes improve build reliability and performance, especially for ARM-based development environments and multi-platform Docker builds.

## QA/ How is this tested?
- [x] Manually

## :rotating_light: Call to action 

This works great on MAC M1 but we need to test this on other machines because we merge it. would be great if you can check it  by running `./publish_docker_images.sh -t develop` and tell us if it does not take much time. It should not take more than few minutes.

## Related Tickets & Documents
closes https://github.com/metr-systems/backlog/issues/4395#issuecomment-3607211272

imenattatra committed:
improve multi-platform build and frontend performance

- Add BUILDPLATFORM and TARGETPLATFORM build arguments and use --platform=$BUILDPLATFORM for the frontend builder stage
to enable native platform builds (improves speed, especially on ARM Macs).
- Remove --network-concurrency 1 from Yarn install, allowing parallel downloads now that native builds are used.
- Minor cleanup and reordering for clarity.

thiagogds reviewed (changes requested):
You need to update the publish script to build using

`docker buildx build --no-cache --build-arg install_groups="main,metr" --platform linux/amd64 -t redash-metr .`

That's because only the buildx has the `BUILDPLATFORM` set automatically

imenattatra committed:
Replace docker build with docker buildx for cross-platform support

This change is necessary because Docker buildx provides enhanced multi-platform build support that docker build alone does not support.

imenattatra commented:
> You need to update the publish script to build using
> 
> `docker buildx build --no-cache --build-arg install_groups="main,metr" --platform linux/amd64 -t redash-metr .`
> 
> That's because only the buildx has the `BUILDPLATFORM` set automatically

You were right! i was not aware that `buildx` is necessary to unlock `BuildKit` full power and possibility to support multi-platform build especially.
